lib/WUI/automata/core.cpp

```diff
@@ -0,0 +1,143 @@
+
+#include <cassert>
+#include <cctype>
+
+using std::nullopt;
+using std::optional;
+
+namespace {
+
+typedef bool (*SpecialCheck)(uint8_t);
+
+bool anything(uint8_t) {
+    return true;
+}
+
+bool whitespace(uint8_t byte) {
+    return isspace(byte);
+}
+
+bool horiz_space(uint8_t byte) {
+    return byte == ' ' or byte == '\t';
+}
+
+bool digit(uint8_t byte) {
+    return isdigit(byte);
+}
+
+SpecialCheck specials[] = {
+    anything,
+    whitespace,
+    horiz_space,
+    digit,
+};
+
+}
+
+namespace automata {
+
+bool Transition::matches(uint8_t byte) const {
+    switch (label_type) {
+    case LabelType::Char:
+        return byte == label;
+    case LabelType::CharNoCase:
+        return tolower(byte) == label;
+    case LabelType::Special:
+        return specials[label](byte);
+    default:
+        assert(0);
+        return false;
+    }
+}
+
+std::optional<Event> Automaton::gen_event(StateIdx old_state, StateIdx new_state, uint8_t payload) const {
+    const State &o = states[old_state];
+    const bool leave = o.emit_leave;
+    const State &n = states[new_state];
+    const bool enter = n.emit_enter;
+
+    if (leave || enter) {
+        return Event {
+            old_state,
+            new_state,
+            leave,
+            enter,
+            payload,
+        };
+    } else {
+        return nullopt;
+    }
+}
+
+std::optional<TransitionResult> Automaton::transition(StateId old, uint8_t byte) const {
+    const State &state = states[old.state];
+    if (state.has_path) {
+        const uint8_t converted = state.path_nocase ? tolower(byte) : byte;
+        const char *path = paths[state.path].value;
+        if (path[old.path] == converted) {
+            if (path[old.path + 1] == '\0') {
+                StateId new_state(old.state + 1);
+                return TransitionResult {
+                    new_state,
+                    gen_event(old.state, new_state.state, byte),
+                    false,
+                };
+            } else {
+                return TransitionResult { old.path_step(), nullopt, false };
+            }
+        }
+    }
+
+    const State &sentinel = states[old.state + 1];
+    for (TransIdx i = state.first_transition; i < sentinel.first_transition; i++) {
+        if (transitions[i].matches(byte)) {
+            const auto new_state = transitions[i].target_state;
+            return TransitionResult { new_state, gen_event(old.state, new_state, byte), transitions[i].fallthrough };
+        }
+    }
+
+    return nullopt;
+}
+
+Execution::~Execution() {}
+
+ExecutionControl Execution::feed(uint8_t byte) {
+    const auto new_state = automaton->transition(current_state, byte);
+
+    if (new_state.has_value()) {
+        current_state = new_state->new_state;
+        if (new_state->emit_event.has_value()) {
+            const auto result = event(*new_state->emit_event);
+            if (result != ExecutionControl::Continue) {
+                return result;
+            }
```

Comment:
I think it might be used in tests (or maybe not).

Anyway, that's kind-of future-proof design. I was thinking about other use cases than just http and if we wanted to eg. parse the multipart forms (which is not that terrible right now, so it definitely isn't on the table right now), there are cases where the upper level may want to just abort the parsing on the semantic level instead of the syntactic one. Like, if the user wants to upload a file but didn't provide the file name, or the file name contains forbidden characters, we want to give up at that point without having to worry about file not being ready when dealing with the payload.